File: java/src/tarehart/rlbot/time/Duration.java
package tarehart.rlbot.time;

public class Duration implements Comparable<Duration> {

    private final long millis;

    private Duration(long millis) {
        this.millis = millis;
    }

    public long toMillis() {
        return millis;
    }

    public static Duration between(GameTime first, GameTime later) {
        return new Duration(later.toMillis() - first.toMillis());
    }

    public static Duration ofMillis(long millis) {
        return new Duration(millis);
    }

    public static Duration ofSeconds(double seconds) {
        return new Duration((long) (seconds * 1000));
    }

    public Duration abs() {
        return  millis < 0 ? new Duration(-millis) : this;
    }

    public double getSeconds() {
        return millis / 1000.0;
    }

    @Override
    public int compareTo(Duration other) {
        return Long.compare(millis, other.millis);
    }

    public Duration plusSeconds(double seconds) {
        return new Duration(millis + (long) (seconds * 1000));
    }
}

File: src/main/java/tarehart/rlbot/time/Duration.kt
package tarehart.rlbot.time

class Duration private constructor(val millis: Long) : Comparable<Duration> {

    val seconds: Double
        get() = millis / 1000.0

    fun abs(): Duration {
        return if (millis < 0) Duration(-millis) else this
    }

    override fun compareTo(other: Duration): Int {
        return java.lang.Long.compare(millis, other.millis)
    }

    fun plusSeconds(seconds: Double): Duration {
        return Duration(millis + (seconds * 1000).toLong())
    }

    operator fun minus(other: Duration): Duration {
        return Duration.ofMillis(millis - other.millis)
    }

    operator fun plus(other: Duration): Duration {
        return Duration.ofMillis(millis + other.millis)
    }

    operator fun times(other: Double): Duration {
        return Duration.ofMillis((millis * other).toLong())
    }

    override fun toString(): String {
        return "$millis ms"
    }


    companion object {
        val ZERO = ofMillis(0)

        fun between(first: GameTime, later: GameTime): Duration {
            return Duration(later.toMillis() - first.toMillis())
        }

        fun ofMillis(millis: Long): Duration {
            return Duration(millis)
        }

        fun ofSeconds(seconds: Double): Duration {
            return Duration((seconds * 1000).toLong())
        }

        fun max(d1: Duration, d2: Duration): Duration {
            return if (d1 > d2) d1 else d2
        }
    }
}
